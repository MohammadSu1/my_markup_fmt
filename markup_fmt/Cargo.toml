[package]
name = "markup_fmt"
version = "0.3.2"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Configurable HTML/Vue/Svelte/Jinja/Twig formatter."
repository = "https://github.com/g-plane/markup_fmt"
license = "MIT"
exclude = ["/tests"]

[dependencies]
css_dataset = "0.3"
itertools = "0.11"
serde = { version = "1.0", optional = true }
tiny_pretty = { version = "0.1", features = ["unicode-width"] }

[dev-dependencies]
insta = { version = "1.34", features = ["glob"] }
toml = "0.8"

[features]
config_serde = ["serde"]
