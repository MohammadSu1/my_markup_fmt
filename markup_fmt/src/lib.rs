mod ast;
pub mod config;
mod ctx;
mod helpers;
mod parser;
mod printer;

pub use crate::parser::Language;
use crate::{
    ctx::Ctx,
    parser::{Parser, SyntaxError},
    printer::DocGen,
};
use config::FormatOptions;
use std::{borrow::Cow, path::Path};
use tiny_pretty::{IndentKind, PrintOptions};

pub fn format_text<F>(
    code: &str,
    language: Language,
    options: &FormatOptions,
    external_formatter: F,
) -> Result<String, SyntaxError>
where
    F: for<'a> Fn(&Path, &'a str) -> Cow<'a, str>,
{
    let mut parser = Parser::new(code, language.clone());
    let ast = parser.parse_root()?;
    let ctx = Ctx {
        language,
        indent_width: options.layout.indent_width,
        options: &options.language,
        external_formatter,
    };
    Ok(tiny_pretty::print(
        &ast.doc(&ctx),
        &PrintOptions {
            indent_kind: if options.layout.use_tabs {
                IndentKind::Tab
            } else {
                IndentKind::Space
            },
            line_break: options.layout.line_break.clone().into(),
            width: options.layout.print_width,
            tab_size: options.layout.indent_width,
        },
    ))
}
