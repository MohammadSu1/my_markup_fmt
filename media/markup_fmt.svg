<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created with Inkscape (http://www.inkscape.org/) -->

<svg
   width="512"
   height="512"
   viewBox="0 0 135.46667 135.46667"
   version="1.1"
   id="svg1"
   inkscape:export-filename="markup_fmt.png"
   inkscape:export-xdpi="48"
   inkscape:export-ydpi="48"
   sodipodi:docname="markup_fmt.svg"
   inkscape:version="1.3.2 (091e20ef0f, 2023-11-25, custom)"
   xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape"
   xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
   xmlns="http://www.w3.org/2000/svg"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:dc="http://purl.org/dc/elements/1.1/">
  <title
     id="title1">Logo of markup_fmt</title>
  <sodipodi:namedview
     id="namedview1"
     pagecolor="#ffffff"
     bordercolor="#000000"
     borderopacity="0.25"
     inkscape:showpageshadow="2"
     inkscape:pageopacity="0.0"
     inkscape:pagecheckerboard="0"
     inkscape:deskcolor="#d1d1d1"
     inkscape:document-units="mm"
     inkscape:zoom="0.82"
     inkscape:cx="163.41463"
     inkscape:cy="226.21951"
     inkscape:window-width="2194"
     inkscape:window-height="1172"
     inkscape:window-x="0"
     inkscape:window-y="0"
     inkscape:window-maximized="1"
     inkscape:current-layer="layer1"
     showgrid="false" />
  <defs
     id="defs1">
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect15"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,36,0,1 @ F,0,0,1,0,36,0,1 @ F,0,0,1,0,36,0,1 @ F,0,0,1,0,36,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect14"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect13"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect12"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect11"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <inkscape:path-effect
       effect="fillet_chamfer"
       id="path-effect10"
       is_visible="true"
       lpeversion="1"
       nodesatellites_param="F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1 @ F,0,0,1,0,0,0,1"
       radius="0"
       unit="px"
       method="auto"
       mode="F"
       chamfer_steps="1"
       flexible="false"
       use_knot_distance="true"
       apply_no_radius="true"
       apply_with_radius="true"
       only_selected="false"
       hide_knots="false" />
    <linearGradient
       id="swatch9">
      <stop
         style="stop-color:#e34c26;stop-opacity:1;"
         offset="0"
         id="stop9" />
    </linearGradient>
    <clipPath
       clipPathUnits="userSpaceOnUse"
       id="clipPath34">
      <rect
         style="fill:none;fill-opacity:1;stroke:#ffffff;stroke-width:0;stroke-linecap:butt;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1"
         id="rect35"
         width="135.46666"
         height="141.45474"
         x="0"
         y="-3.7097557e-15"
         transform="matrix(1,0,0.20355859,0.97906277,0,0)" />
    </clipPath>
    <clipPath
       clipPathUnits="userSpaceOnUse"
       id="clipPath35">
      <rect
         style="fill:none;fill-opacity:1;stroke:#ffffff;stroke-width:0;stroke-linecap:butt;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1"
         id="rect36"
         width="135.46666"
         height="141.45474"
         x="0"
         y="-3.7097557e-15"
         transform="matrix(1,0,0.20355859,0.97906277,0,0)" />
    </clipPath>
  </defs>
  <g
     inkscape:label="layer1"
     inkscape:groupmode="layer"
     id="layer1"
     style="display:inline">
    <rect
       style="display:inline;mix-blend-mode:normal;fill:#e34c26;fill-opacity:1;stroke:#000000;stroke-width:0;stroke-linecap:butt;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1"
       id="rect18"
       width="91.223244"
       height="159.31955"
       x="-0.57189626"
       y="-17.548292"
       transform="matrix(1,0,-0.20791169,0.9781476,0,0)"
       mask="none"
       clip-path="url(#clipPath35)" />
    <rect
       style="display:inline;mix-blend-mode:normal;fill:#f06529;fill-opacity:1;stroke:#000000;stroke-width:0;stroke-linecap:butt;stroke-linejoin:round;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1"
       id="rect18-5"
       width="88.254929"
       height="154.13544"
       x="90.480286"
       y="-12.5835"
       transform="matrix(1,0,-0.20791169,0.9781476,0,0)"
       mask="none"
       clip-path="url(#clipPath34)" />
    <g
       id="g20"
       transform="matrix(1.7779259,0,-0.37790985,1.7779259,-8.2753109,-37.11192)"
       style="display:inline;mix-blend-mode:normal;fill:#ffffff;fill-opacity:1;stroke:#ffffff;stroke-width:0;stroke-dasharray:none;stroke-opacity:1">
      <path
         style="font-weight:900;font-size:11.2889px;font-family:Inter;-inkscape-font-specification:'Inter Heavy';stroke-linejoin:round"
         d="m 18.953625,52.278083 v -1.929255 l 5.964156,-2.392277 v 2.028474 l -3.726219,1.289845 v 0.07717 l 3.726219,1.289846 v 2.017449 z m 7.474487,3.136418 v -8.213116 h 3.516757 l 0.738629,2.568666 q 0.08819,0.303169 0.192926,0.788239 0.110243,0.48507 0.220486,1.04731 0.110243,0.56224 0.20395,1.102431 0.09371,-0.540191 0.20395,-1.102431 0.110243,-0.56224 0.214974,-1.04731 0.110243,-0.48507 0.192926,-0.788239 l 0.727605,-2.568666 h 3.522269 v 8.213116 h -2.248961 v -3.175003 q 0,-0.314193 0.01654,-0.854385 0.01654,-0.545703 0.03307,-1.179601 0.02205,-0.639411 0.03307,-1.240236 -0.132292,0.622874 -0.286633,1.25126 -0.148828,0.628386 -0.292144,1.157553 -0.137804,0.529167 -0.237023,0.865409 l -0.953603,3.175003 h -1.852086 l -0.970139,-3.175003 q -0.09922,-0.319705 -0.237023,-0.826824 -0.137804,-0.507118 -0.286632,-1.102431 -0.148829,-0.600826 -0.281121,-1.207163 0.01103,0.584289 0.02756,1.196138 0.01654,0.606338 0.03307,1.124481 0.01654,0.51263 0.01654,0.815799 v 3.175003 z m 12.545673,0.09922 q -0.887457,0 -1.449698,-0.440973 -0.556728,-0.440972 -0.556728,-1.355991 0,-0.677995 0.297657,-1.091407 0.303169,-0.413412 0.815799,-0.617362 0.512631,-0.209462 1.146529,-0.253559 0.771703,-0.05512 1.09692,-0.159853 0.325217,-0.110243 0.325217,-0.380339 v -0.02205 q 0,-0.237023 -0.187413,-0.363803 -0.187414,-0.132291 -0.507119,-0.132291 -0.314193,0 -0.540191,0.137804 -0.220487,0.132291 -0.275608,0.413411 l -1.995402,-0.198437 q 0.08819,-0.518143 0.429949,-0.942579 0.341754,-0.424437 0.942579,-0.677996 0.606337,-0.253559 1.471746,-0.253559 0.876433,0 1.521356,0.270096 0.650435,0.264583 1.003213,0.722092 0.352778,0.45751 0.352778,1.025262 v 4.222313 H 40.792797 V 54.54358 h -0.0441 q -0.28112,0.518143 -0.727605,0.744142 -0.446485,0.225998 -1.04731,0.225998 z m 0.716581,-1.400088 q 0.374826,0 0.677995,-0.231511 0.303169,-0.23151 0.303169,-0.650435 v -0.507118 q -0.341754,0.132292 -0.881946,0.220486 -0.385851,0.06063 -0.56224,0.231511 -0.176389,0.170877 -0.176389,0.4079 0,0.259071 0.176389,0.396875 0.181902,0.132292 0.463022,0.132292 z m 4.299483,1.300869 v -6.162593 h 2.149741 v 1.179602 h 0.06615 q 0.165365,-0.655947 0.529168,-0.953603 0.369314,-0.303169 0.859896,-0.303169 0.28112,0 0.56224,0.08819 v 1.885159 q -0.165364,-0.06063 -0.407899,-0.08268 -0.237023,-0.02756 -0.4079,-0.02756 -0.490582,0 -0.815799,0.319706 -0.319706,0.319705 -0.319706,0.848872 v 3.208076 z m 5.071184,0 v -8.213116 h 2.215887 v 4.123095 h 0.07717 l 1.488282,-2.072572 h 2.491496 l -2.09462,2.700958 2.149742,3.461635 h -2.524569 l -1.256772,-2.17179 -0.33073,0.418924 v 1.752866 z m 8.995845,0.07717 q -0.650434,0 -1.135504,-0.297656 -0.48507,-0.297657 -0.749654,-0.821312 -0.264584,-0.523655 -0.264584,-1.196138 v -3.924657 h 2.215888 v 3.461636 q 0,0.457509 0.231511,0.722092 0.23151,0.259072 0.650434,0.259072 0.4079,0 0.650435,-0.259072 0.242535,-0.264583 0.242535,-0.722092 v -3.461636 h 2.215888 v 6.162593 h -2.078084 l -0.03858,-1.389064 q -0.225998,0.628386 -0.68902,1.04731 -0.463021,0.418924 -1.25126,0.418924 z m 5.181428,2.226912 v -8.466675 h 2.193839 v 1.080383 h 0.05512 q 0.176389,-0.474045 0.61185,-0.815799 0.43546,-0.341754 1.118968,-0.341754 0.61185,0 1.157553,0.325217 0.551216,0.325218 0.898482,1.025262 0.347266,0.694532 0.347266,1.8135 0,1.063846 -0.330729,1.76389 -0.33073,0.700045 -0.881946,1.047311 -0.551216,0.341753 -1.212675,0.341753 -0.650434,0 -1.085895,-0.30868 -0.429948,-0.314193 -0.622874,-0.782727 h -0.03307 v 3.318319 z m 3.14193,-3.924656 q 0.457509,0 0.705556,-0.385851 0.253559,-0.385852 0.253559,-1.069359 0,-0.68902 -0.253559,-1.074871 -0.248047,-0.391363 -0.705556,-0.391363 -0.451997,0 -0.716581,0.391363 -0.264583,0.385851 -0.264583,1.074871 0,0.666971 0.264583,1.063846 0.264584,0.391364 0.716581,0.391364 z"
         id="text19"
         aria-label="&lt;Markup" />
      <path
         style="font-weight:900;font-size:11.2889px;font-family:Inter;-inkscape-font-specification:'Inter Heavy';stroke-linejoin:round"
         d="m 58.825928,69.504974 v -8.213116 h 5.776742 v 1.796964 h -3.54983 v 1.852085 h 3.197052 v 1.752867 H 61.05284 v 2.8112 z m 6.498834,0 v -6.162593 h 2.022963 l 0.110243,1.262285 q 0.264583,-0.722093 0.744141,-1.030774 0.48507,-0.308681 1.04731,-0.308681 0.584289,0 0.964628,0.336242 0.380339,0.336241 0.633898,1.152041 0.292144,-0.8158 0.84336,-1.152041 0.556728,-0.336242 1.17409,-0.336242 0.881945,0 1.433161,0.567752 0.551216,0.567753 0.551216,1.515844 v 4.156167 h -2.215888 v -3.593927 q 0,-0.402387 -0.203949,-0.628386 -0.20395,-0.231511 -0.545704,-0.231511 -0.341754,0 -0.540192,0.231511 -0.198437,0.225999 -0.198437,0.628386 v 3.593927 h -2.116669 v -3.593927 q 0,-0.402387 -0.20395,-0.628386 -0.20395,-0.231511 -0.545704,-0.231511 -0.347266,0 -0.545703,0.231511 -0.192926,0.225999 -0.192926,0.628386 v 3.593927 z m 14.254442,-6.162593 v 1.609551 h -1.058335 v 2.574178 q 0,0.385851 0.418924,0.385851 0.09922,0 0.286633,-0.02205 0.192925,-0.02756 0.253559,-0.04961 l 0.264584,1.581989 q -0.380339,0.104731 -0.716581,0.137804 -0.336242,0.03858 -0.633898,0.03858 -2.089108,0 -2.089108,-1.868622 v -2.778127 h -0.782727 v -1.609551 h 0.782727 v -1.466234 h 2.215887 v 1.466234 z m 4.927869,-2.436374 -2.645836,9.833691 h -1.874134 l 2.645836,-9.833691 z m 7.011466,6.228739 -5.964156,2.381253 v -2.01745 l 3.737244,-1.289845 v -0.07717 l -3.737244,-1.289845 v -2.028474 l 5.964156,2.392277 z"
         id="text20"
         aria-label="Fmt/&gt;" />
    </g>
  </g>
  <metadata
     id="metadata1">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <cc:license
           rdf:resource="http://creativecommons.org/licenses/by-nc-nd/4.0/" />
        <dc:creator>
          <cc:Agent>
            <dc:title>Pig Fang</dc:title>
          </cc:Agent>
        </dc:creator>
        <dc:title>Logo of markup_fmt</dc:title>
        <dc:subject>
          <rdf:Bag>
            <rdf:li>markup_fmt</rdf:li>
          </rdf:Bag>
        </dc:subject>
      </cc:Work>
      <cc:License
         rdf:about="http://creativecommons.org/licenses/by-nc-nd/4.0/">
        <cc:permits
           rdf:resource="http://creativecommons.org/ns#Reproduction" />
        <cc:permits
           rdf:resource="http://creativecommons.org/ns#Distribution" />
        <cc:requires
           rdf:resource="http://creativecommons.org/ns#Notice" />
        <cc:requires
           rdf:resource="http://creativecommons.org/ns#Attribution" />
        <cc:prohibits
           rdf:resource="http://creativecommons.org/ns#CommercialUse" />
      </cc:License>
    </rdf:RDF>
  </metadata>
</svg>
