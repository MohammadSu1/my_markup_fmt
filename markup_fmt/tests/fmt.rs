use insta::{assert_snapshot, glob, Settings};
use markup_fmt::{format_text, Language};
use std::fs;

#[test]
fn fmt_snapshot() {
    glob!("fmt/**/*.{html,vue,svelte}", |path| {
        let input = fs::read_to_string(path).unwrap();
        let language = match path.extension().unwrap().to_str().unwrap() {
            "html" => Language::Html,
            "vue" => Language::Vue,
            "svelte" => Language::Svelte,
            _ => unreachable!("unknown file extension"),
        };

        let options = fs::read_to_string(path.with_file_name("config.toml"))
            .map(|config_file| toml::from_str(&config_file).unwrap())
            .unwrap_or_default();

        let output = format_text(&input, language.clone(), &options, |_, code, _| {
            Ok::<_, ()>(code.into())
        })
        .map_err(|err| format!("failed to format '{}': {:?}", path.display(), err))
        .unwrap();
        let regression_format = format_text(&output, language, &options, |_, code, _| {
            Ok::<_, ()>(code.into())
        })
        .unwrap();
        assert_eq!(
            output,
            regression_format,
            "'{}' format is unstable",
            path.display()
        );

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(name, output);
        });
    });
}
