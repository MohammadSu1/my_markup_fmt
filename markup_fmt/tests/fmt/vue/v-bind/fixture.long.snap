---
source: markup_fmt/tests/fmt.rs
---
<template>
  <button v-bind:type=""></button>
  <button v-bind:type.sync=""></button>
  <button v-bind:type=""></button>
  <button v-bind:type.sync=""></button>
</template>

